### PR TITLE
Fix  issue#1037 with updated weights

### PR DESCRIPTION
**Pull request by HamidShojanazeri**
## Description

This PR fixes issue #1037. Following the discussions from PR #1038, that we were not able to load the pre-trained weights for image classification examples on different gpu device (i.e., map_locaiton=self.device/cuda+id), this happened because these weights were serialized using older version of serialization code, and torchvision team is upgrading weights in the master/nightly release now, you can read more [on this issue](https://github.com/pytorch/vision/issues/3767).

This PR, is using the updated weights from vision master branch and using vgg16 as vgg11 was not updated yet.

Fixes #(issue)
#1037 
## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## steps to test
```
git clone https://github.com/pytorch/serve.git
git checkout issue_1037_weight_update
python ./ts_scripts/install_dependencies.py --environment=dev --cuda=cu110 # or 102
python ts_scripts/install_from_src.py
```

You can test any of the vision models by following the README, here we use vgg16 example.

```
wget https://download.pytorch.org/models/vgg16-397923af.pth
torch-model-archiver --model-name vgg16 --version 1.0 --model-file ./examples/image_classifier/vgg_16/model.py --serialized-file vgg16-397923af.pth --handler ./examples/image_classifier/vgg_16/vgg_handler.py --extra-files ./examples/image_classifier/index_to_name.json
```

```
mkdir model_store

mv vgg16.mar model_store/

```
In a seprate temrinal:

`torchserve --start --model-store model_store`

Then in a seprate terminal:

```
curl -X POST "localhost:8081/models?model_name=vgg16&url=vgg16.mar&batch_size=1&max_batch_delay=5000&initial_workers=4&synchronous=true"
curl http://127.0.0.1:8080/predictions/vgg16 -T ./examples/image_classifier/kitten.jpg
```
In a separate terminal you want to watch the gpu usage and make sure per each worker there is only one process on all the gpus.

`watch -n01 nvidia-smi`

### Logs of the example run
[run_log.txt](https://github.com/pytorch/serve/files/6437520/run_log.txt)

- UT/IT execution results

- Logs
All working examples are tested on P2.8xlarge model both for eager and traced models.
regression test logs are added. 
[regression_logs.txt](https://github.com/pytorch/serve/files/6437361/regression_logs.txt)

Screen Shot 2021-05-04 at 11 13 11 AM

## Checklist:

- [ ] Have you added tests that prove your fix is effective or that this feature works?
- [ ] New and existing unit tests pass locally with these changes?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [X] Have you made corresponding changes to the documentation?

### CODE REVIEW

**Comment by msaroufim** [comment]
@HamidShojanazeri looking at your test plan screenshot it's not clear to me that this indeed achieves the desired behavior. It's better to have commands and logs in text so it's easier for me to repro since there were small minor issues like no directory called vgg11 in the repo and some missing steps like starting the torch serve model

I couldn't repro the model working fine locally

I tried all the flavors of CUDA_EXPORT_DEVICE you listed but all of them gave me the 500 error below

Also why are we testing Alexnet, my understanding is this issue is specific to vgg16 and vgg11

```
gh pr checkout 1052
python ./ts_scripts/install_dependencies.py --environment=dev --cuda=cu110
python ./ts_scripts/install_from_src.py
wget https://download.pytorch.org/models/vgg16-397923af.pth
mv vgg16.mar model_store
torch-model-archiver --model-name vgg16 --version 1.0 --serialized-file vgg16-397923af.pth --extra-files examples/image_classifier/index_to_name.json --handler examples/image_classifier/vgg_16/vgg_handler.py
torchserve --start --model-store model_store --models vgg16=vgg16.mar
```

In a separate window

`curl -X POST "localhost:8081/models?model_name=vgg16&url=vgg16.mar&batch_size=1&max_batch_delay=5000&initial_workers=4&synchronous=true"`

Full error log here https://gist.github.com/msaroufim/f6f525d9d6f17a28c8d794fad1437728

```
ubuntu@ip-172-31-1-60:~/serve$ curl -X POST "localhost:8081/models?model_name=vgg16&url=vgg16.mar&batch_size=1&max_batch_delay=5000&initial_workers=4&synchronous=true"
{
  "code": 500,
  "type": "InternalServerException",
  "message": "Failed to start workers"
}
```

File: .circleci/scripts/linux_test_sanity.sh [hunk 1]
```diff
@@ -2,7 +2,7 @@
```
Comment (msaroufim) [inline .circleci/scripts/linux_test_sanity.sh:2]: How come we're updating all the vision models? I thought this issue was specific to vgg. Same question for alexnet, resnet etc..

File: test/postman/inference_data.json [hunk 2]
```diff
@@ -151,38 +151,38 @@
         "tolerance":5
```
Comment (msaroufim) [inline test/postman/inference_data.json:151]: Should this be checked into the repo?

File: test/pytest/test_model_archiver.py [hunk 3]
```diff
@@ -6,7 +6,7 @@
 import json
 import test_utils
 
-MODEL_SFILE_NAME = 'resnet18-5c106cde.pth'
```
Comment (msaroufim) [inline test/pytest/test_model_archiver.py:9]: This feels like it's gonna break again as soon they release updated pretrained models. Is there a convention the team uses where the names increase monotonically so you can justs pick the latest available models automatically?

**Review by msaroufim — CHANGES_REQUESTED** [review]

File: .circleci/scripts/linux_test_sanity.sh (hunk 1 above)
Comment (HamidShojanazeri) [reply .circleci/scripts/linux_test_sanity.sh:2]: @msaroufim This issue exist for five vision models, that they are now updated in the torchvision/master. You can read more about it [here](https://github.com/pytorch/vision/issues/3767#issuecomment-831825167)

File: test/pytest/test_model_archiver.py (hunk 3 above)
Comment (HamidShojanazeri) [reply test/pytest/test_model_archiver.py:9]: @msaroufim I am not aware of any naming convention as per I know, but in general this won't break soon, as map_location now is in full support.

**Comment by HamidShojanazeri** [comment]
> @HamidShojanazeri looking at your test plan screenshot it's not clear to me that this indeed achieves the desired behavior. It's better to have commands and logs in text so it's easier for me to repro since there were small minor issues like no directory called vgg11 in the repo and some missing steps like starting the torch serve model
> 
> I couldn't repro the model working locally
> 
> I tried all the flavors of CUDA_EXPORT_DEVICE you listed but all of them gave me the 500 error below
> 
> So step by step instructions for a test plan would be very helpful
> 
> Also why are we testing Alexnet, my understanding is this issue is specific to vgg16 and vgg11
> 
> ```
> gh pr checkout 1052
> python ./ts_scripts/install_dependencies.py --environment=dev --cuda=cu110
> python ./ts_scripts/install_from_src.py
> wget https://download.pytorch.org/models/vgg16-397923af.pth
> mv vgg16.mar model_store
> torch-model-archiver --model-name vgg16 --version 1.0 --serialized-file vgg16-397923af.pth --extra-files examples/image_classifier/index_to_name.json --handler examples/image_classifier/vgg_16/vgg_handler.py
> torchserve --start --model-store model_store --models vgg16=vgg16.mar
> ```
> 
> In a separate window
> 
> `curl -X POST "localhost:8081/models?model_name=vgg16&url=vgg16.mar&batch_size=1&max_batch_delay=5000&initial_workers=4&synchronous=true"`
> 
> Full error log here https://gist.github.com/msaroufim/f6f525d9d6f17a28c8d794fad1437728
> 
> ```
> ubuntu@ip-172-31-1-60:~/serve$ curl -X POST "localhost:8081/models?model_name=vgg16&url=vgg16.mar&batch_size=1&max_batch_delay=5000&initial_workers=4&synchronous=true"
> {
>   "code": 500,
>   "type": "InternalServerException",
>   "message": "Failed to start workers"
> }
> ```
> 
> I also tried following the instructions in the vgg16 readme but same issue where it failed to start workers

I will try to repro, the error but seems the PR should be # 1053 seems you checked out to 1052. Sorry for the steps, it was meant to be reviewed by team who was working on the issue. The regression tests log with new mar files has been added to the logs.

**Comment by msaroufim** [comment]
> I will try to repro, the error but seems the PR should be # 1053 seems you checked out to 1052. Sorry for the steps, I will update them, it was meant to be reviewed by team who was working on the issue. The regression tests log with new mar files has been added to the logs.

Apologies that was a typo in copy paste, the run was correctly getting #1053 and I just reran now to double check

**Comment by HamidShojanazeri** [comment]
> > I will try to repro, the error but seems the PR should be # 1053 seems you checked out to 1052. Sorry for the steps, I will update them, it was meant to be reviewed by team who was working on the issue. The regression tests log with new mar files has been added to the logs.
> 
> Apologies that was a typo in copy paste, the run was correctly getting #1053 and I just reran now to double check
> 
> EDIT: The issue went away when I instead called `curl -X POST "localhost:8081/models?model_name=vgg16&url=vgg16.mar&batch_size=1&max_batch_delay=5000`
@msaroufim Thanks for the updates, I could not repro, and added the steps above and logs of the run. From the steps you shared seems once model was registered with torchserve --start then it was tried again to register with Curl command.

**Comment by msaroufim** [comment]
When I try to register with any number of workers it gives a 500 error failed to register

So I tried having 0 workers and then scaling with the API 

And now it just hangs when I run the inference `curl http://127.0.0.1:8080/predictions/vgg16 -T ./examples/image_classifier/kitten.jpg`

Logs here https://gist.github.com/msaroufim/146952c55f3f56767e2f37625e794e73

**Comment by HamidShojanazeri** [comment]
> When I try to register with any number of workers it gives a 500 error failed to register
> 
> So I tried having 0 workers and then scaling with the API
> 
> And now it just hangs when I run the inference `curl http://127.0.0.1:8080/predictions/vgg16 -T ./examples/image_classifier/kitten.jpg`
> 
> Logs here https://gist.github.com/msaroufim/146952c55f3f56767e2f37625e794e73

@msaroufim are you first registering the model using torchserve --start --model-store model_store --models vgg16=vgg16.mar?

**Comment by msaroufim** [comment]
EDIT 2: It worked when I setup a fresh environment

Screen Shot 2021-05-06 at 4 35 16 PM


Screen Shot 2021-05-06 at 4 31 08 PM

**Comment by msaroufim** [comment]
Just one final point I tried running on master with a new venv `torch-model-archiver --model-name vgg11 --version 1.0 --serialized-file vgg11-bbd30ac9.pth --extra-files examples/image_classifier/index_to_name.json --handler examples/image_classifier/vgg_11/vgg_handler.py
torchserve --start --model-store model_store --models vgg11=vgg11.mar`

And I didn't see any issues with duplicate PIDs 
Screen Shot 2021-05-06 at 4 48 38 PM

So just double checking whether it is needed to change `device` to `location` or whether #1037 is fixed by only updating the weights

**Review by msaroufim — APPROVED** [review]

**Review by lxning — APPROVED** [review]

**Pull request merged**